File: src/main/java/com/example/demo/service/ProduitService.java
package com.example.demo.service;

import com.example.demo.entity.Produit;
import java.util.List;

public interface ProduitService {
    Produit creerProduit(Produit produit);
    Produit obtenirProduitParId(Long id);
    List<Produit> obtenirTousLesProduits();
    Produit mettreAJourProduit(Long id, Produit produit);
    void supprimerProduit(Long id);
    List<Produit> rechercherProduitsParNom(String nom);
}

File: projectmanagerbackend/gestion-projets-frontend/src/app/services/produit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Produit {
  id?: number;
  nom: string;
  prix: number;
  description?: string;
}

@Injectable({
  providedIn: 'root'
})
export class ProduitService {

  private apiUrl = 'http://localhost:8080/api/produits';

  constructor(private http: HttpClient) {}

  getProduits(): Observable<Produit[]> {
    return this.http.get<Produit[]>(this.apiUrl);
  }

  getProduitById(id: number): Observable<Produit> {
    return this.http.get<Produit>(`${this.apiUrl}/${id}`);
  }

  creerProduit(produit: Produit): Observable<Produit> {
    return this.http.post<Produit>(this.apiUrl, produit);
  }

  // ✅ Méthode pour mettre à jour un produit (PUT)
  mettreAJourProduit(id: number, produit: Produit): Observable<Produit> {
    return this.http.put<Produit>(`${this.apiUrl}/${id}`, produit);
  }

  supprimerProduit(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  rechercherParNom(nom: string): Observable<Produit[]> {
    const params = new HttpParams().set('nom', nom);
    return this.http.get<Produit[]>(`${this.apiUrl}/recherche`, { params });
  }
}
